isort/api.py

```diff
@@ -297,70 +302,76 @@ def sort_file(
     - **write_to_stdout**: If `True`, write to stdout instead of the input file.
     - ****config_kwargs**: Any config modifications.
     """
-    with io.File.read(filename) as source_file:
-        actual_file_path = file_path or source_file.path
-        config = _config(path=actual_file_path, config=config, **config_kwargs)
-        changed: bool = False
-        try:
-            if write_to_stdout:
-                changed = sort_stream(
-                    input_stream=source_file.stream,
-                    output_stream=sys.stdout,
-                    config=config,
-                    file_path=actual_file_path,
-                    disregard_skip=disregard_skip,
-                    extension=extension,
-                )
-            else:
-                tmp_file = source_file.path.with_suffix(source_file.path.suffix + ".isorted")
-                try:
-                    with tmp_file.open(
-                        "w", encoding=source_file.encoding, newline=""
-                    ) as output_stream:
-                        shutil.copymode(filename, tmp_file)
-                        changed = sort_stream(
-                            input_stream=source_file.stream,
-                            output_stream=output_stream,
-                            config=config,
-                            file_path=actual_file_path,
-                            disregard_skip=disregard_skip,
-                            extension=extension,
-                        )
-                    if changed:
-                        if show_diff or ask_to_apply:
-                            source_file.stream.seek(0)
-                            with tmp_file.open(
-                                encoding=source_file.encoding, newline=""
-                            ) as tmp_out:
-                                show_unified_diff(
-                                    file_input=source_file.stream.read(),
-                                    file_output=tmp_out.read(),
-                                    file_path=actual_file_path,
-                                    output=None if show_diff is True else cast(TextIO, show_diff),
-                                    color_output=config.color_output,
-                                )
-                                if show_diff or (
-                                    ask_to_apply
-                                    and not ask_whether_to_apply_changes_to_file(
-                                        str(source_file.path)
+    try:
```

Comment:
I think the try except should be elsewhere (probably on the call of this function, or , better on ` encoding, _ = tokenize.detect_encoding(buffer.readline)` in `io.py` so it's more localized and the change affect less code than that.

Reply:
@Pierre-Sassoulas I have also added a try clause on `encoding, _ = tokenize.detect_encoding` in `isort/io.py`. The reason why I also put a try catch block here as well as `isort/main.py` is that I'm trying to throw a warning for each file with invalid encoding(if verbose flag is set) and not an error even if at least one file has valid encoding. The error is only for the case when all the files processed have invalid encodings. So I needed a way to propagate this `invalid encodings exceptions` to `isort/main.py` where I am finally deciding if to throw an error or not. `isort/io.py` and `isort/api.py` are only called per file so I couldn't come up with a way to do it there.